set.mm

```diff
@@ -33738,6 +33760,16 @@ technically classes despite morally (and provably) being sets, like ` 1 `
       MLN $.
   $}
 
+  ${
+    $d A x $.  $d B x $.  $d ph x $.
+    eqelssd.1 $e |- ( ph -> A C_ B ) $.
+    eqelssd.2 $e |- ( ( ph /\ x e. B ) -> x e. A ) $.
+    $( Equality deduction from subclass relationship and membership.
+       (Contributed by AV, 21-Aug-2022.) $)
+    eqelssd $p |- ( ph -> A = B ) $=
+      ( cv wcel ex ssrdv eqssd ) ACDEABDCABGZDHLCHFIJK $.
```

Comment:
When I see this statement and am told that it shortens many proofs, I'm not sure what to make of it.  This could be a sign that said proofs are not very clear: probably more readable proofs would have proved ph -> A C_ B and ph -> B C_ A and from these two subproofs, ph -> A = B. But if we shorten them using eqelssd, the deficiency of these proofs will be harder to notice... I had made such a remark on the discussion group, and I realize that I am not offering a solution here... What do others think (@digama0, @avekens, @savask, @jkingdon) ?

Reply:
1. eqelssd is a valid shortening according to the rules on the conventions page.  The new theorem adds 9 extra lines to set.mm, by contrast the shortenings in the main section only save already 13 lines.  The advantage is much greater, if one takes mathboxes into account.  So from a formal standpoint one can hardly argue against it.  Actually, eqssd is part of it's proof, it seems that the target expression is frequently constructed from ( ( ph /\ x e. B ) -> x e. A ) as a starting point.  
2. Your discomfort seems to be based on a sort of pedagogical aspect.  Some applications may hide, for example, symmetric aspects of the proof, that were otherwise better visible.  If one was to present the proof to an audience and pinpoint the essential steps, one might get lost when the proof is built from lots of adhoc theorems, not clearly focussing on a single thought.  I think this is a valid observation, and some visiting Metamath for the first time may even be repelled.  To some extent the concept of the Metamath language reminds me of Assembler in my profession: Extremely fast results on the machine, hard to grasp for a human being.  I like to use the term 'frogs-eye' perspective for this effect
3. If you really want to address the 'machine-focussed' style, a first step in my eyes is breaking long proofs down into lemmas, each focussing on a single aspect of the proof.  Here you have full control over the layout of a proof, and a wise choice could express the 'birds-eye view' properly. This could be a starting point for your intention.
4. When looking into how programming languages have evolved, I'd say, you need on top of the simple replacement scheme, that Metamath offers, an extension that allows you to express concepts.   But off the top of my head, I can't come up with something useful.  If this hadn't been solved already in some other proof program, a meta-Metamath language will take time and experience to develop.
5. First step in such a situation is (my experience) analysis of the current situation using statistics for example.  Even if the current rules are sometimes to the detriment of human readers, this might still be a neglectable portion of the whole database.

Reply:
Since `( ph /\ x e. B ) -> x e. A` is just an `ex` and an `ssrdv` away from `B C_ A`, at least to my eyes `eqelssd` is a perfectly valid way of expressing the idea that showing the subclass relationship in both directions implies equality. In fact, it might be more clear than something with an explicit `B C_ A` if it avoids a few steps which are just housekeeping rather than statements about the classes in question.

So I don't see a conflict in this case between "shorter" and "more readable".

Reply:
There are 66 proofs using this pattern (wherefore ~eqelssd is used for them now). Are all of these proofs badly designed? Maybe two or three of them should be analyzed in more detail, to find out why this pattern is used (and if it can be avoided). Without concrete examples, a judgement about ~eqelssd is difficult.

Reply:
I had a look at ~dvaddf which is proofed using eqelssd now:
```
67 37,39,41,42,44,45,46,56,65,66 dvaddbr  $p |- ( ( ph /\ x e. X )
                                          -> x ( S _D ( F oF + G ) ) ( ( ( S _D F ) ` x ) + ( ( S _D G ) ` x ) ) )
68 reldv         $p |- Rel ( S _D ( F oF + G ) )
69 68 releldmi   $p |- ( x ( S _D ( F oF + G ) ) ( ( ( S _D F ) ` x ) + ( ( S_D G ) ` x ) ) -> x e. dom ( S _D ( F oF + G ) ) )
70 67,69 syl     $p |- ( ( ph /\ x e. X ) -> x e. dom ( S _D ( F oF + G ) ) )
71 35,70 eqelssd  $p |- ( ph -> dom ( S _D ( F oF + G ) ) = X )
```

Here, it is ~releldmi which requires this pattern. and I see no easy/more elegant way to prove that ` X C_ dom ( S _D ( F oF + G ) ) ` in this case.

Reply:
Ok. This is more of an open question about a general phenomenon where shortening with an ad-hoc lemma may sometimes hide more conceptual and larger shortenings.
In any case, this is non-blocking for a merge.

Reply:
I see: it may happen that one inclusion is more naturally proved as A C_ B while the reverse inclusion is more naturally proved using a dummy variable.